File: new-fund-service/updateFunds.js
// ==================== CONFIGURATION ====================
require('dotenv').config();
const mongoose = require('mongoose');
const fetch = require('node-fetch');
const fs = require('fs');
const Fund = require('./models/Fund');

const API_BASE_URL = "https://api.mfapi.in/mf";

// --- Stability Settings (from Python script) ---
const CONCURRENCY_LIMIT = 10;
const RETRY_COUNT = 4;
const RETRY_DELAY_SECONDS = 3;

// --- Filtering & Output ---
const DAYS_THRESHOLD = 14;
const OUTPUT_FILENAME = "active_funds_backup.json";

// --- MONGODB SETTINGS ---
const MONGO_URI = process.env.MONGO_URI;
// =======================================================


/**
 * Parses a DD-MM-YYYY date string into a Date object.
 * @param {string} dateStr - The date string to parse.
 * @returns {Date | null}
 */
function parseDDMMYYYY(dateStr) {
    if (!dateStr) return null;
    try {
        const [day, month, year] = dateStr.split('-').map(Number);
        return new Date(year, month - 1, day);
    } catch (e) {
        return null;
    }
}

/**
 * Fetches details for a single fund with robust retries.
 * @param {object} fund - The fund object from the master list.
 * @returns {Promise<object | null>}
 */
async function fetchFundDetails(fund) {
    const schemeCode = fund.schemeCode;
    // Ensure schemeCode is a valid number
    if (!schemeCode || isNaN(Number(schemeCode))) return null;

    const url = `${API_BASE_URL}/${schemeCode}`;

    for (let attempt = 0; attempt < RETRY_COUNT; attempt++) {
        try {
            const response = await fetch(url, { timeout: 30000 });
            if (response.status === 200) {
                const data = await response.json();
                const navData = data?.data;

                if (!Array.isArray(navData) || navData.length === 0) return null;

                const latestNav = navData[0];
                const navDate = parseDDMMYYYY(latestNav?.date);

                if (!navDate) return null;

                const dateThreshold = new Date();
                dateThreshold.setDate(dateThreshold.getDate() - DAYS_THRESHOLD);

                if (navDate >= dateThreshold) {
                    return {
                        code: schemeCode,
                        name: data?.meta?.scheme_name || "N/A",
                        nav: latestNav?.nav,
                        date: latestNav?.date,
                        last_updated_on: new Date()
                    };
                }
                return null; // Fund is not active
            } else {
                console.warn(`\n⚠️ Status ${response.status} for fund ${schemeCode} (Attempt ${attempt + 1}/${RETRY_COUNT})`);
            }
        } catch (e) {
            console.warn(`\n⚠️ Error fetching fund ${schemeCode} (Attempt ${attempt + 1}/${RETRY_COUNT}): ${e.message}`);
        }

        if (attempt < RETRY_COUNT - 1) {
            await new Promise(res => setTimeout(res, RETRY_DELAY_SECONDS * 1000 * (attempt + 1)));
        }
    }
    console.error(`\n❌ Failed to fetch fund ${schemeCode} after ${RETRY_COUNT} retries.`);
    return null;
}


/**
 * Connect to MongoDB
 */
async function connectDB() {
    try {
        await mongoose.connect(MONGO_URI, {
            dbName: 'test'
        });
        console.log('📦 Connected to MongoDB');
    } catch (error) {
        console.error('❌ MongoDB connection error:', error);
        throw error;
    }
}

/**
 * Main function to run the entire process.
 */
async function main() {
    console.log("🚀 Starting the mutual fund updater...");
    if (!MONGO_URI) {
        console.error("❌ CRITICAL ERROR: MONGO_URI is not set in your .env file.");
        return;
    }

    try {
        // Connect to MongoDB
        await connectDB();
    } catch (error) {
        console.error("❌ Failed to connect to MongoDB. Exiting.");
        return;
    }

    let allFunds;

    try {
        console.log("📡 Fetching the master list of all funds...");
        const response = await fetch(API_BASE_URL);
        if (!response.ok) throw new Error(`HTTP Error: ${response.status}`);
        allFunds = await response.json();
    } catch (e) {
        console.error(`❌ Critical Error: Could not fetch master fund list. ${e}`);
        return;
    }

    const totalFunds = allFunds.length;
    console.log(`✅ Found ${totalFunds} total funds to check.`);

    const activeFunds = [];
    let progressCounter = 0;

    for (let i = 0; i < totalFunds; i += CONCURRENCY_LIMIT) {
        const batch = allFunds.slice(i, i + CONCURRENCY_LIMIT);
        const promises = batch.map(fund => fetchFundDetails(fund));
        const results = await Promise.all(promises);

        results.forEach(result => {
            progressCounter++;
            // Enhanced validation to ensure we only include valid funds
            if (result && 
                result.code && 
                !isNaN(Number(result.code)) && 
                result.name && 
                result.nav && 
                result.date) {
                activeFunds.push(result);
            }
        });
        process.stdout.write(`⚙️  Processing: ${progressCounter}/${totalFunds} funds | Found: ${activeFunds.length} active\r`);
    }

    console.log("\n\n✅ Processing complete.");
    console.log("=" * 30);
    console.log(`📊 FINAL RESULTS:`);
    console.log(`   - ✅ Active Funds:    ${activeFunds.length}`);
    console.log(`   - ❌ Inactive/Failed: ${totalFunds - activeFunds.length}`);
    console.log("=" * 30);

    if (activeFunds.length > 0) {
        try {
            console.log("\n💾 Connected to MongoDB...");

            console.log("🗑️  Deleting all old documents from 'active_funds' collection...");
            const { deletedCount } = await Fund.deleteMany({});
            console.log(`   -> ${deletedCount} documents deleted.`);

            console.log(`✍️  Inserting ${activeFunds.length} new active funds into MongoDB...`);
            await Fund.insertMany(activeFunds, { 
                ordered: false,  // Continue processing even if some documents fail
                collection: 'active_funds'
            }).catch(err => {
                console.warn("Warning: Some documents failed to insert:", err.message);
                // Continue execution even if there are some failures
            });
            console.log("   -> Successfully inserted valid documents.");

            console.log("\n🎉 MongoDB update complete.");
        } catch (e) {
            console.error(`\n❌ CRITICAL ERROR: Could not update MongoDB. Error: ${e}`);
        } finally {
            await mongoose.connection.close();
            console.log("🔒 MongoDB connection closed.");
        }
    }

    console.log(`\n💾 Saving local backup to '${OUTPUT_FILENAME}'...`);
    fs.writeFileSync(OUTPUT_FILENAME, JSON.stringify(activeFunds, null, 2));
    console.log("🎉 All done!");
}


// --- RUN SCRIPT ---
main();

File: python-fund-updater/update_funds.py
import asyncio
import aiohttp
import json
import os
from datetime import datetime, timedelta
from typing import Optional, Dict, Any
from motor.motor_asyncio import AsyncIOMotorClient
from dotenv import load_dotenv # Import the dotenv library

# --- LOAD ENVIRONMENT VARIABLES ---
load_dotenv() # This will load the .env file from the current directory

# ==================== CONFIGURATION ====================
API_BASE_URL = "https://api.mfapi.in/mf"

# --- Stability Settings ---
CONCURRENCY_LIMIT = 10
RETRY_COUNT = 4
RETRY_DELAY_SECONDS = 3

# --- Filtering & Output ---
DAYS_THRESHOLD = 14
OUTPUT_FILENAME = "active_funds_backup.json"

# --- MONGODB SETTINGS ---
MONGO_URI = os.getenv("MONGO_URI")
DB_NAME = "test"
COLLECTION_NAME = "py_active_fund"
# =======================================================


def parse_ddmmyyyy(date_str: str) -> Optional[datetime]:
    """Parses a DD-MM-YYYY date string into a datetime object."""
    try:
        return datetime.strptime(date_str, "%d-%m-%Y")
    except (ValueError, TypeError):
        return None


async def fetch_fund_details(
    session: aiohttp.ClientSession,
    fund: Dict[str, Any],
    semaphore: asyncio.Semaphore
) -> Optional[Dict[str, Any]]:
    """Fetches details for a single fund with robust retries and error logging."""
    scheme_code = fund.get("schemeCode")
    if not scheme_code:
        return None

    url = f"{API_BASE_URL}/{scheme_code}"

    async with semaphore:
        await asyncio.sleep(0.05)

        for attempt in range(RETRY_COUNT):
            try:
                async with session.get(url, timeout=30) as response:
                    if response.status == 200:
                        data = await response.json()
                        nav_data = data.get("data")
                        
                        if not nav_data or not isinstance(nav_data, list) or not nav_data:
                            return None

                        latest_nav = nav_data[0]
                        nav_date = parse_ddmmyyyy(latest_nav.get("date"))
                        
                        if not nav_date:
                            return None

                        if nav_date >= datetime.now() - timedelta(days=DAYS_THRESHOLD):
                            return {
                                "code": scheme_code,
                                "name": data.get("meta", {}).get("scheme_name", "N/A"),
                                "nav": latest_nav.get("nav"),
                                "date": latest_nav.get("date"),
                                "last_updated_on": datetime.utcnow()
                            }
                        return None
                    
                    else:
                        print(f"\n⚠ Status {response.status} for fund {scheme_code} (Attempt {attempt+1}/{RETRY_COUNT})")

            except Exception as e:
                error_message = str(e).split('[Errno')[0].strip()
                print(f"\n⚠ Error fetching fund {scheme_code} (Attempt {attempt+1}/{RETRY_COUNT}): {error_message}")
            
            if attempt < RETRY_COUNT - 1:
                await asyncio.sleep(RETRY_DELAY_SECONDS * (attempt + 1))

    print(f"\n❌ Error fetching fund {scheme_code}: Failed permanently after {RETRY_COUNT} retries.")
    return None


async def main():
    """Main function to run the entire filtering and database update process."""
    print("🚀 Starting the mutual fund updater...")
    
    if not MONGO_URI:
        print("❌ CRITICAL ERROR: MONGO_URI not found. Please create a .env file and set it.")
        return

    async with aiohttp.ClientSession() as session:
        print("📡 Fetching the master list of all funds...")
        try:
            async with session.get(API_BASE_URL) as response:
                response.raise_for_status()
                all_funds = await response.json()
        except Exception as e:
            print(f"❌ Critical Error: Could not fetch master fund list. {e}")
            return
            
    total_funds = len(all_funds)
    print(f"✅ Found {total_funds} total funds to check.")

    semaphore = asyncio.Semaphore(CONCURRENCY_LIMIT)
    progress_counter = 0
    
    async with aiohttp.ClientSession() as session:
        tasks = [fetch_fund_details(session, fund, semaphore) for fund in all_funds]
        
        active_funds = []
        for task in asyncio.as_completed(tasks):
            result = await task
            progress_counter += 1
            if result:
                active_funds.append(result)
            
            print(f"⚙️  Processing: {progress_counter}/{total_funds} funds | Found: {len(active_funds)} active", end='\r')

    # --- FINAL COUNTS ---
    active_count = len(active_funds)
    inactive_count = total_funds - active_count
    
    print("\n\n✅ Processing complete.")
    print("="*30)
    print(f"📊 FINAL RESULTS:")
    print(f"    - ✅ Active Funds:    {active_count}")
    print(f"    - ❌ Inactive/Failed: {inactive_count}")
    print("="*30)

    # --- SAVE TO MONGODB ---
    if active_funds:
        client = None
        try:
            print(f"\n💾 Connecting to MongoDB...")
            client = AsyncIOMotorClient(MONGO_URI)
            db = client[DB_NAME]
            collection = db[COLLECTION_NAME]

            print(f"🗑️  Deleting all old documents from '{COLLECTION_NAME}'...")
            delete_result = await collection.delete_many({})
            print(f"   -> {delete_result.deleted_count} documents deleted.")

            print(f"✍️  Inserting {active_count} new active funds into MongoDB...")
            await collection.insert_many(active_funds)
            print(f"   -> Successfully inserted documents.")
            
            print("\n🎉 MongoDB update complete.")

        except Exception as e:
            print(f"\n❌ CRITICAL ERROR: Could not update MongoDB. Error: {e}")
        finally:
            if client:
                client.close()
                print("🔒 MongoDB connection closed.")
    else:
        print("\n⚠️ No active funds found to insert into the database.")

    # --- SAVE LOCAL BACKUP ---
    print(f"\n💾 Saving local backup to '{OUTPUT_FILENAME}'...")
    with open(OUTPUT_FILENAME, "w", encoding="utf-8") as f:
        def default(o):
            if isinstance(o, datetime):
                return o.isoformat()
        json.dump(active_funds, f, indent=2, ensure_ascii=False, default=default)
        
    print("🎉 All done!")


if __name__ == "__main__":
    asyncio.run(main())
